# Short form of answers/ideas for common CS coding interview questions
___
___

# Riddles of sorts
___
___

## Egg Drop

### Problem Statement

A tower has 100 floors. You've been given two eggs. The eggs are strong enough that they can be dropped from a particular floor in the tower without breaking. You've been tasked to find the highest floor an egg can be dropped without breaking, in as few drops as possible. If an egg is dropped from above its target floor it will break. If it is dropped from that floor or below, it will be intact and you can test drop the egg again on another floor.

Show algorithmically how you would go about doing this in as few drops as possible. (Your answer should be a number of the fewest drops needed for testing 2 eggs on 100 floors)

### Solution

##### Linear Search (Simplest)

- Drop #1 from every 10th floor.
- When #1 breaks, drop #2 each of last 9 floors till #2 breaks.

Best Case: 1st floor -> 1 + 1 = 2 drops total

Worst Case: 99th floor -> 10 + 9 = 19 drops total

##### Triangle series (Optimal)

Generalize the problem to **n** floors.  If there are 1000 floors, then drops for #1 is the primary contributer.  So, need to find solution to where number of drops for #1 + #2 is constant for all cases of a given building.  

Consider:
- Drop #1 from **x**
- If #1 breaks, worst case is #2 takes **x-1**.  **total = x+(x-1)**
- Keep trying #1 from every x+(x-1).

Triangle series:

$$x + (x-1) + (x-2) + (x-3) + ... + 1$$

which equals x(x+1)/2.  Solve for x knowing total number of floors.  So,

$$x(x+1)/2 = n$$

Solve with numpy:  convert to zero on one side, then:
$$0 = 0.5x^2 +0.5x -n  $$  

In [3]:
import numpy as np
n=100
np.roots([0.5,0.5,-n])

array([-14.6509717,  13.6509717])

Positive root is 13.651.

So, first floor to try is 14.

Worst case is 14.

___
___
# Common coding Qs

### Fibonacci series

Each element of the Fibonacci series is the sum of the two proceeding elements. The simplest is the series 1, 1, 2, 3, 4, 5, etc.

## Write a function that computes the Nth fibonacci number

In [11]:
#Brute force, looping. (Simplest)
def fib(n):
    
    a,b = 1,1
    for i in range(n-1):
        a,b = b,a+b
    return a

print "brute force: ",fib(7)

# Using recursion (slowest)
def fibR(n):
    if n==1 or n==2:
        return 1
    return fib(n-1)+fib(n-2)

print "recursion: ",fibR(7)

## Using memoization (Fastest).  Memoize the simple loop/brute force func.
def memoize(fn, arg):
    memo = {}
    if arg not in memo:
        memo[arg] = fn(arg)
    return memo[arg]
fibm = memoize(fib,7)
print "memoization: ",fibm

brute force:  13
recursion:  13
memoization:  13


## max profits

### Problem

Given a list of stock prices, where the index is the time stamp, find the greatest profit that can be made by a single purchase and sale.

For example, if you were given the list of stock prices:

prices = [12,11,15,3,10]

Then your function would return the maximum possible profit, which would be 7 (buying at 3 and selling at 10).

### Solution.  

#### Brute Force.  O(N^2) time, O(1) space.

Loop over every element and find max value after it, keeping track of max profit.

#### Greedy Algorith.  O(N) time, O(1) space.

Go through list, tracking:
    - min price so far
    - max profit so far (ie. max(current_price-min, max_profit) )
    
Corner cases:
- less than 2 prices in list
- prices keep going down.

Fixes:
- set initial min price to 0th
- set initial max price to 1st - 0th